### ISSUE TITLE
Docker Image

### ISSUE THREAD

**Issue by AddoSolutions** [nodemailer/wildduck#71]
Hey guys, just wanted to see if there was anyone already working on a docker image for this project.  I have a pretty long history with Docker, and have dockerized a few other projects so I can certainly help significantly with this if need be.

The one really big help I have found is your install script.  It doesn't appear to be fully documented, but it only seems to work on Ubuntu Xenial flavor?  With that image, I was able to get it to install fairly well into an image, but there are several important tweaks for the install that would increase it's stability/standardize it's installation.

This is a VERY cool project, and glad that someone is FINALLY working on a replacement for the horrid file based mail systems written in a modern language.  I actually started doing it myself but ran out of time to finish (some of my progress is [here](https://github.com/AddoSolutions/unsuckish-mail) if it's of any help, though this repo is FAR more advanced).  

Let me know if I can help on this project.  I am pretty seasoned in NodeJS, MongoDB, and Docker -Redis being my weak point.

Nick

**Comment by AddoSolutions** [comment]
For anyone wanting to play with docker, you can get it about as far as I did with:
do:

```shell
# On the host
docker run --rm -it --link mongodb --link redis -p 8080:8080 -p 993:993 ubuntu:xenial bash

# In the newly created temporary container
apt-get update; \
    apt-get -y install lsof wget gnupg dnsutils curl sudo git; \
    wget https://raw.githubusercontent.com/nodemailer/wildduck/master/setup/install.sh; \
    chmod +x install.sh; \
    ./install.sh ; \
    echo 'mongo="mongodb://mongodb:27017/wildduck"
    redis="redis://redis:6379/3"' > /etc/wildduck/dbs.toml; \
    cd /opt/wildduck; \
    /usr/bin/node server.js --config="/etc/wildduck/wildduck.toml"
```

**Comment by andris9** [comment]
Actually there is nothing Ubuntu specific, for example I run the server in a custom CoreOS based system. The install script uses apt for installing the dependencies (MongoDB and such), and it uses systemd to run the services (thus the requirement for Ubuntu 16, not 14 which does not have systemd). If you manage Ubuntu and Redis separately and start the service process as a docker command then there's nothing you'd need from Ubuntu.

**Comment by AddoSolutions** [comment]
Very cool! Yeah it does a really nice Job, I’m just trying to figure out how to best dockerize it. 

From a strategic perspective is Docker something you guys are interested in formally supporting or am I going to be better off doing it separately?

**Comment by andris9** [comment]
If you are willing to maintain the docker stuff then sure, create a PR with the docker files and I'll merge.

**Comment by larstobi** [comment]
@AddoSolutions Did you make it? :-)

**Comment by AddoSolutions** [comment]
I actually didn't get to finish this project on my end.  Moving off Gmail but that is a big frog to swallow lol.

The config above should get you quite far though, you just need to make sure to add the volumes in the right place.

Something like:

```
FROM ubuntu:xenial
WORKDIR /opt/wildduck
RUN apt-get update; \
    apt-get -y install lsof wget gnupg dnsutils curl sudo git; \
    wget https://raw.githubusercontent.com/nodemailer/wildduck/master/setup/install.sh; \
    chmod +x install.sh; \
    ./install.sh ; \
    echo 'mongo="mongodb://mongodb:27017/wildduck"
    redis="redis://redis:6379/3"' > /etc/wildduck/dbs.toml
CMD /usr/bin/node server.js --config="/etc/wildduck/wildduck.toml
```

**Comment by rqelibari** [comment]
Hey there!

We have dockerized Wildduck mail server ([astzweig/wildduck](https://hub.docker.com/r/astzweig/wildduck)) and did release version [1.0.0](https://github.com/astzweig/docker-wildduck) today on GitHub. It contains a full fledged email server, but no Wildduck webmail, as we will release that in it's own docker container shortly.

## Highlights
* The container can be configured using just a few different environment variables and using the API of course. Everything else can be induced by other values. If you need more control, do not hesitate to tweak the configuration files directly (mountable).
* We developed a python program ([cocof](https://pypi.org/project/cocof/)) specifically for the
   purpose to modify YAML, TOML and JSON configuration files from CLI scripts and can therefor
   configure the system in short, concise scripts, that are far less error-prone than `sed`-based scripts.
* Most important part of our work was to actually understand how the system components relate to
  each other and map that to a good script structure.

Hope it saves you guys some time. And as always: feedback is very welcome! We sure put our best efforts into it - at least our best effort besides our day-to-day work. Enjoy :-)

**Comment by kerberjg** [comment]
@rqelibari `astzweig/wildduck` is behind on several releases of Wildduck, the docker image is still stuck on 1.1, while the latest version of Wildduck is actually 1.15

The best solution would be for Wildduck to have its own Dockerfiles in the repo, as well as autobuild configured on Docker Hub, in order for the project to be self-reliant, without the risk of a lack of updated builds which would depends entirely on external maintainers.

We're currently to integrate Wildduck as a part of our dockerized infrastructure over at @mekomidev, and we'd gladly make a pull request once everything is done on our sides and if @andris9 agrees on this.

**Comment by unixfox** [comment]
@kerberjg Based on the [Dockerfile](https://github.com/astzweig/docker-wildduck/blob/master/Dockerfile) the image is based on commit https://github.com/nodemailer/wildduck/commit/bda1cbd878e5c02dabea54886059e1fd54514054 which is the v1.15.1 of wildduck, so I don't think `astzweig/wildduck` is actually that outdated.

**Comment by andris9** [comment]
I've been really bad with changelogs (to say I haven't kept one) but comparing the latest version against astzweig/wildduck the differences in WildDuck core are mostly in filter handling:
* if a filter moves messages to a folder and that folder is deleted then the filter is deleted as well
* if there is afilter that tries to store messages to an unexisting folder then the message is stored to INBOX instead
(previous versions returned 450 error for such a message)

Additionally the authentication API endpoint was changed to return user specific access token so mail clients would be able to use user token instead of root token when making requests on behalf of the user.

So in large part the astzweig/wildduck version should work still fine.

**Comment by rqelibari** [comment]
@kerberjg Generally a good idea. Bumping the container automatically would be a nice thing, maybe through some kind of continuous integration service.

Additionally: I will prepare a new release manually today.

**Comment by manalejandro** [comment]
Something like:
```
version: "2"
services:
    wildduck:
        build: ./wildduck
        ports:
            - "25:25"
            - "465:465"
            - "993:993"
        volumes:
             - ./wildduck/config:/wildduck/config
             - ./wildduck-mta/config:/wildduck-mta/config
             - ./haraka/config:/haraka/config
        depends_on:
            - redis
            - mongo

    redis:
        image: redis:alpine
        volumes:
          - ./redis:/data
        expose:
          - 6379
    
    mongo:
        image: mongo
        volumes:
          - ./mongodb:/data/db
        expose:
          - 27017

    webmail:
        build: ./webmail
        ports:
          - "3000:3000"
        volumes:
          - ./webmail/config:/webmail/config
        depends_on:
          - redis
          - mongo
          - wildduck

```

:+1:

**Comment by kerberjg** [comment]
@unixfox if we're talking about the *docker image* `astzweig/wildduck`, its release tags do not match actual Wildduck releases, which is misleading. If a user is pulling an image named `wildduck`, they're expecting the version tag to match `wildduck`'s version, and this is not the case.

@rqelibari My proposal would include configuring this repo for Docker Hub's autobuilds, it's the new official CI service for Docker which also provides builds for alternative arch's out of the box

**Comment by rqelibari** [comment]
## @kerberjg 
### Release Number
Matching Wildduck releases is a bit of work semantically speaking. `astzweig/wildduck` depends on multiple software all with their own release cycles and numbers e.g. like the Wildduck Haraka plugin or the Wildduck ZoneMTA plugin.

### Automated builds
An ideal solution would be a CI service, that would create a new Docker container release when any of the software changes and only if the container remains functional. For the latter a kind of test suite would be necessary. Is that possible with Docker Hub's autobuild feature?

## Next step
Before continuing the discussion I propose @andris9 shares his vision on how a Docker service shall be integrated in regard to this repository.

**Comment by andris9** [comment]
Actually I do not have a preference as I do not use Docker myself. If you'd like me to set up some Github webhooks to trigger once anything new is pushed to start some kind of a build then I could do that. I could also accept a Dockerfile and docker-compose files (or whatever else is needed) directly into WildDuck repo - but I would not be interested in maintaining these.

**Comment by manalejandro** [comment]
Well, i have been deploying with docker-compose and haraka + wildduck + webmail since some months and is very powerfull and stable, think in support that please. Future is `docker` :nerd_face:

**Comment by mhf-ir** [comment]
Any update on official docker image ?

**Comment by rqelibari** [comment]
@mhf-ir Did you not read the conversation? I quote @andris9:

> If you'd like me to set up some Github webhooks to trigger once anything new is pushed to start some kind of a build then I could do that. I could also accept a Dockerfile and docker-compose files (or whatever else is needed) directly into WildDuck repo - but I would not be interested in maintaining these.

Are you willing to integrate e.g. [astzweig/docker-wildduck](https://github.com/astzweig/docker-wildduck) and maintain it?

**Comment by manalejandro** [comment]
I just published my containers on https://hub.docker.com/u/manalejandro :whale:

**Comment by unixfox** [comment]
@manalejandro sorry for this offtopic (I'm going to remove this comment after you reply to it).
Your gitea is misconfigured, I can't access it from IPv6 only host. As you can see the IPv6 record associated with gitea.hatthieves.es is invalid: 
![image](https://user-images.githubusercontent.com/4016501/79080555-72a5d300-7d05-11ea-9a02-bf210005a9b9.png)

**Comment by manalejandro** [comment]
> @manalejandro sorry for this offtopic (I'm going to remove this comment after you reply to it).
> Your gitea is misconfigured, I can't access it from an IPv6 only host. As you can see the IPv6 record associated with gitea.hatthieves.es is invalid:
> ![image](https://user-images.githubusercontent.com/4016501/79080555-72a5d300-7d05-11ea-9a02-bf210005a9b9.png)

sorry will solve it now, thanks (will delete it too)

**Comment by almereyda** [comment]
Thanks for the interesting chat in my inbox!

How did you do this IPv6 validation?

(Will delete this post from the interface, too.)

**Comment by manalejandro** [comment]
> Thanks for the interesting chat in my inbox! How did you do this IPv6 validation? (Will delete this post from the interface, too.)

First we didn't have ipv6, requested it from the provider and then fought to configure it, now everything works fine with ipv6 including Docker, thanks for the notice :smile: (will be deleted too)

**Comment by rqelibari** [comment]
FYI: Since the merge of PR #180, wildduck has it's own and pure Docker image at: https://hub.docker.com/u/nodemailer

**Comment by manalejandro** [comment]
> FYI: Since the merge of PR #180, wildduck has it's own and pure Docker image at: https://hub.docker.com/u/nodemailer

Thanks i'm probably late, but it's not a full mail server with `haraka` :wink: maybe ideas for the docker project :hammer:

**Comment by louis-lau** [comment]
A full mail Serv image is already linked above: https://hub.docker.com/r/astzweig/wildduck

In my opinion it's nice to have the containers separate as you can then scale and distribute them individually :).

**Comment by louis-lau** [comment]
Hey all! I just released [WildDuck: dockerized](https://github.com/nodemailer/wildduck-dockerized), which _is_ a complete mail server including haraka, zonemta, etc.

It's different from @rqelibari's excellent image in that all services are deployed in seperate containers, the docker way! It's also an official installation method for wildduck.

Try it out and let me know if you come across any issues.

**Comment by manalejandro** [comment]
I wrote one article in my blog about deploy wildduck-dockerized

https://manalejandro.com/~/Art%C3%ADculos/stack-de-correo-electr%C3%B3nico-completo-con-node-js-haraka-wildduck-y-docker/

Thanks louis

**Comment by louis-lau** [comment]
Thanks @manalejandro, but you missed the part about the setup script which will put your domain name in the right places, and copies the default config. The way you've explained it right there will not work, as there's no `config` folder.

You don't have to clone the repo, the setup image does everything for you.

**Comment by louis-lau** [comment]
I see you've updated your guide, but it's still not entirely correct. You don't clone the repo, and you don't execute the setup.sh script on the host. You only run the setup docker image, like it says in the readme:
```
$ mkdir wildduck-dockerized
$ cd wildduck-dockerized
$ docker run --rm -v "${PWD}:/wildduck-dockerized" nodemailer/wildduck-dockerized-setup domainname [hostname]
```

No need to clone, or even have git installed.

Thanks for writing about it on your blog :)

**Comment by manalejandro** [comment]
Louis I have put the two ways, the one to clone and the other you like :wink: thanks a lot

**Comment by louis-lau** [comment]
The one with the clone does not work ;)

**Comment by manalejandro** [comment]
Yes, but i have deployed the stack in one Pi4 with your docker-compose.yml and my config folder, it works for me :smile:

**Comment by louis-lau** [comment]
Sure! But if you clone like you say in your blog post, you don't have a config folder. You also say to execute the setup.sh script on the host, this does not work.

Yes, you can copy the docker-compose, .env, config folder, and change all needed values. But, the setup image does this in one click. You also don't describe this process in your blog post. The way it's written now, it will not work. Try it yourself on a vps if you want.

**Comment by manalejandro** [comment]
Ok, i will put the details :smile:
